### DISCUSSION TITLE
Graph visualization

### DISCUSSION THREAD

**Discussion by Orbifold** [observablehq/framework#1010] [Q&A]
I specialize in all things graphs and graph visualization. Use quite a bit of [Ogma](https://doc.linkurious.com/ogma/latest/) (by Linkurious) and yFiles (by [yWorks](https://www.yworks.com)).

I have tried to integrate this type of dataviz into the framework but all attempts failed. Is there more to it than `npm install`, creating a component and `display()`? The error is simply `unable to load the module`. Tried all module types.

**Comment by Fil** [comment]
npm install would install the libraries on the "server-side"—whereas, I guess, you probably want them on the "client-side" (to run them interactively in the browser). You could try something like this:

````
```js
import * as ogma from "npm:ogma"
```
````

the "ogma" npm module  I saw seems unrelated to graph visualization, so I have not been able to go much further, but happy to help if you can share more details.

**Reply by Orbifold** [reply]
My message contains a link to Ogma, what d'you mean it's unrelated to graph visualization? 

No, Ogma is not on the public npm repo and you can't use `npm:` since this will try to fetch it from some CDN.
Precisely my question in fact: how to include packages if they are not public and if the standard `npm install` ain't doin it?

**Reply by Fil** [reply]
I searched on npm and found https://www.npmjs.com/package/ogma which was clearly unrelated.

I don't have a straightforward answer because it all depends on how you're supposed to use the resources that the module provides. If it provides a function that does computation, you might do the computations (server-side) in a data loader, and use the computed data in your page. If it's something that needs to run on the page (client-side), it probably needs some kind of "translation" to the browser (usually with rollup).

**Comment by mbostock** [comment]
> how to include packages if they are not public and if the standard npm install ain't doin it?

Currently, you’d need to download the source files into your project and import them as a local ES module, e.g.,

```js
import * as ogma from "./ogma.js";
```

You can also import from a URL if it’s hosted somewhere else (but this isn’t recommended because you’ll have an external dependency on this server in your built site, which isn’t great for performance or security):

```js
import * as ogma from "https://example.com/ogma.js";
```

We’re planning on adding support for importing from `node_modules` in #360 which would let you say:

```js
import * as ogma from "ogma";
```

Please upvote #360 if you’re interested in this feature!
